controller/array_action/array_mediator_ds8k.py

```diff
@@ -0,0 +1,402 @@
+POOL_CAP_ALLOCATED_REAL_EXTS_ON_ESE = 'real_capacity_allocated_on_ese'
+POOL_CAP_ALLOCATED_VIRT_EXTS_ON_ESE = 'virtual_capacity_allocated_on_ese'
+
+# flashCopy
+FC_SOURCE_VOLUME_ID = 'sourcevolume'
+FC_TARGET_VOLUME_ID = 'targetvolume'
+FC_STATUS = 'state'
+FC_IS_PERSISTENT = 'persistent'  # enabled|disabled
+FC_IS_RECORDING = 'recording'  # enabled|multinc|disabled|None
+FC_IS_BGCOPY = 'backgroundcopy'  # enabled|disabled
+FC_ID = 'id'
+
+# pprc
+PPRC_SOURCE_VOLUME_ID = 'source_volume'
+PPRC_TARGET_VOLUME_ID = 'target_volume'
+# PPRC_SOURCE_ESS_NAME = 'system'
+PPRC_SOURCE_SYSTEM_ID = 'source_system'
+PPRC_TARGET_SYSTEM_ID = 'target_system'
+PPRC_STATUS = 'state'
+PPRC_REMOTE_COPY_TYPE = 'type'  # metromirror|globalcopy|unknown
+
+# ioport
+IOPORT_NAME = 'id'
+IOPORT_WWPN = 'wwpn'
+# IOPORT_WWNN = ''
+IOPORT_PORT_SPEED = 'speed'
+IOPORT_STATUS = 'state'
+IOPORT_STATUS_ONLINE = 'online'
+IOPORT_LOCATION = 'loc'
+IOPORT_ENCLOSURE_NUMBER = 'io_enclosure'
+
+# host
+HOST_ID = 'name'
+HOST_NAME = 'name'
+HOST_STATE = 'state'
+HOST_TYPE = 'hosttype'  # VMWare
+HOST_VOLUME_MAPPINGS = 'mappings_briefs'
+HOST_VOLUME_MAPPING_VOLUME_ID = 'volume_id'
+HOST_VOLUME_MAPPING_LUN_ID = 'lunid'
+HOST_PORTS = 'host_ports_briefs'
+HOST_ADDRESS_DISCOVERY = 'addrdiscovery'
+
+# host port
+HOST_PORT_WWPN = 'wwpn'
+HOST_PORT_HOST_ID = 'host'
+HOST_PORT_IOPORTS = 'login_ports'
+HOST_PORT_STATE = 'state'
+HOST_PORT_TYPE = 'hosttype'
+
+# host mapping
+HOST_MAPPING_LUN_ID = 'lunid'
+HOST_MAPPING_VOLUME_ID = 'volume'
+
+# user
+USER_NAME = 'name'
+USER_STATE = 'state'
+USER_GROUP = 'group'
+
+# lss
+LSS_ID = 'id'
+LSS_TYPE = 'type'
+LSS_GROUP = 'group'
+
+# io enclosure
+IOENCLOSURE_ID = 'id'
+IOENCLOSURE_NAME = 'name'
+IOENCLOSURE_STATE = 'state'
+
+# node
+NODE_ID = 'id'
+NODE_STATUS = 'state'
+
+# marray
+MARRAY_ID = 'id'
+MARRAY_DISK_CLASS = 'disk_class'
+MARRAY_POOL_ID = 'pool'
+
+# response keys
+RES_STATUS = 'status'
+RES_CODE = 'code'
+RES_MSG = 'message'
+
+# response values
+RES_SUCCESS_STAUTS = ('ok', 'successful', )
+
+# response error codes
+INVALID_CREDENTIALS = 'BE7A002D'
+
+
+def parse_version(bundle):
+    """
+    Parse ds8k version number from bundle.
+
+    rule is:
+    87.51.34.0 => 7.5.1
+    87.50.21.1 => 7.5.0
+    88.0.151.0 => 8.0.0
+    """
+
+    v1, v2, _ = bundle.split('.', 2)
+    v2 = '{0:0{1}}'.format(int(v2), 2)
+
+    return '.'.join([v1[-1], v2[0], v2[1]])
+
+
+def get_volume_id_from_scsi_identifier(scsi_id):
+    return scsi_id[-4:]
+
+
+# shorten volume name to 16 characters if it is too long.
+def shorten_volume_name(name):
+    if len(name) <= 16:
+        return name
+    return base58.b58encode(sha1(name.encode()).digest()).decode()[-16:]
+
+
+class DS8KArrayMediator(ArrayMediator):
+    SUPPORTED_FROM_VERSION = '7.5.1'
+
+    @classproperty
+    def array_type(self):
+        return 'DS8K'
+
+    @classproperty
+    def port(self):
+        return 8452
+
+    @classproperty
+    def max_vol_name_length(self):
+        return 63
```

Comment:
where is this number coming from? shouldn't it be 16?

Reply:
it was 16, but with the help of the shorten_volume_name method, we can support long volume name, but it should not be very long, so I set the limitation just same with svc and xiv.

Reply:
I think the meaning of this field is the actual final limit in the storage (for generic name cutting logic).
If we are going to remove the name cutting, we should remove this field altogether.

Regardless, I don't think the svc/xiv limitations should affect ds8k limitations.

Reply:
1. If we remove this field, how do we know the name length limit and raise the error?
2. Yes, but I just what to be consistent.

Reply:
1. If we remove this field, how do we know the name length limit and raise the error?
2. Yes, but I what to be consistent.

Reply:
1. If the name length limit is not per storage, it should be somewhere else (this class is for storage-specific details)
2. As long as we use this field here, it should not necessarily be consistent (by design), since values here represent  differences between storage types.

Reply:
No, the name length limit is per storage, but ds8k is an exception. Because the limit is too short (16), we have to workaround it. But even if we can workaround it, we should not allow the user to input a very long name.
But how long is "very long"? To be consistent, I choose the svc limit (63).

Reply:
I see. so you chose some number that happen to be equal to the svc limit.
If this number does not come directly from the storage limitation, it should be at least commented (like you did with `max_connections`)

Reply:
good idea, added the comment.